### ISSUE TITLE
Defaults: map gd to CTRL-]

### ISSUE THREAD

**Issue by gpanders** [neovim/neovim#28476]
The default mapping for "jump to tag" is `<C-]>`, but this is awkward or difficult to type on some keyboards.

Nvim's LSP client sets `tagfunc` by default which enables jumping to a definition with `<C-]>`. `gd` is a very popular custom mapping for jumping to a definition, so mapping `gd` to `<C-]>` would essentially provide that mapping for free.

We could either do this universally, or only when an LSP buffer is attached. One downside of doing it universally is that if an LSP buffer is _not_ attached, and there are no valid tags files, then we lose the builtin `gd` behavior (which is, frankly, not very good, but some people may rely on it).

Note that we do not simply map `gd` to `vim.lsp.buf.definition()` by default because using `'tagfunc'` provides a lot of other useful functionality (opening a definition in a split, searching for identifiers with `:ts`, etc.).

**EDIT:** This is really only useful if mapping `gd` to `<C-]>` also gets us `<C-W>gd` "for free", which I'm not sure if it does.

**Comment by altermo** [comment]
We could also do `<C-]>` when either `'tagfunc'` is defined or when `len(tagfiles())` is over 0, and otherwise `gd`.
<details><summary>Example keymap</summary>

```lua
vim.keymap.set('n', 'gd', function()
    if vim.o.tagfunc ~= '' or #vim.fn.tagfiles() > 0 then
        return '<C-]>'
    end
    return 'gd'
end, { expr = true })
```

</details> 

Or we also could do that if `<C-]>` errors with `E426: Tag not found` or `E433: No tags file` the fallback to `gd`, though this may require som C code changes.

**Comment by mfussenegger** [comment]
Note that we currently also have the unmapped:

- `vim.lsp.buf.declaration()`
- `vim.lsp.buf.implementation()`
- `vim.lsp.buf.type_definition()`

all of which also fit into the `gd` / `gD` area. My vote would go towards using the `gd` and `gD` to cover these by default instead of having two ways for go-to-definition.

**Comment by bluss** [comment]
Consider doing this for help and its impact on help? This would increase the usability of help even though that kind of major change will take a long time to establish itself. It's a breaking change there too, and fundamental docs (the help landing page and `|bars|`) would be updated to mention it in that case.

**Comment by clason** [comment]
> Consider doing this for help and its impact on help?

Help already has `K` mapped to it.

**Comment by gpanders** [comment]
The original justification for this suggestion was that it was awkward or difficult to use `<C-]>` on some keyboards, though that appears to not actually be the case? (cc @dundargoc). In that case we can drop the original proposal and repurpose this issue for establishing other defaults that @mfussenegger suggested.

**Comment by gegoune** [comment]
> difficult to use <C-]> on some keyboards, though that appears to not actually be the case?

I am using corne keyboard and it is somewhat inconvenient indeed.
(I like staying as close to defaults as possible and was looking forward for this proposal to materialise.)

**Comment by dundargoc** [comment]
The context is that is that `<C-]>` did not work for me. I had initially
assumed this was a problem with the `]` character and that there was a problem
interpreting the combination needed to write `]` (which on swedish mac is
Option+9). The reason for this assumption was because e.g. `<C-i>` and `<C-o>`
worked as documented.

**This is completely incorrect.** It turns out `<C-]>` *does not* mean `Ctrl + ]`,
it means byte 29 (unsure about the correct terminology) as described in
https://ss64.com/ascii.html. In layman's terms, it will be `Ctrl` + the key
that is on the same location as `]` in an american keyboard. On Swedish
keyboards, which I use, `<C-]>` is actually a combination of `Ctrl` + `¨`. On
French keyboards, it is `Ctrl` + `$`. On Cyrillic keyboards it is `Ctrl` + `ъ`.

This phenomenon is not isolated to `]`, I just never noticed the other cases as
Swedish and American keyboards happens to coincide for other characters. But
for Cyrillic keyboards for example, `<C-o>` is actually a combination of
`CTRL`+`щ` and `<C-i>` is `CTRL`+`ш`.

**Comment by justinmk** [comment]
> In layman's terms, it will be `Ctrl` + the key
> that is on the same location as `]` in an american keyboard. On Swedish
> keyboards, which I use, `<C-]>` is actually a combination of `Ctrl` + `¨`. On
> French keyboards, it is `Ctrl` + `$`. On Cyrillic keyboards it is `Ctrl` + `ъ`.

Seems worth mentioning at `:help ctrl-]`. That lets us use `gd` for something else (as discussed above).

**Comment by bluss** [comment]
I can confirm on a macos 13.x install (Swedish keyboard) that Ctrl-] (typed using Ctrl-Option-9) has no effect in Neovim. Not in Terminal, not in iTerm2 default and not in iTerm2 with CSI u protocol enabled (looks like it has the same effect as typing 9 by itself, i.e with 9 active pending the next command).

**Comment by gpanders** [comment]
Closing per https://github.com/neovim/neovim/issues/28476#issuecomment-2074885532. As mentioned in that comment, the original justification for this proposal does not seem to be accurate.

**Comment by justinmk** [comment]
Should we use this to track https://github.com/neovim/neovim/issues/28476#issuecomment-2074250729 though?

**Comment by gpanders** [comment]
> Should we pivote this to track [#28476 (comment)](https://github.com/neovim/neovim/issues/28476#issuecomment-2074250729) though?

With https://github.com/neovim/neovim/pull/28650 merged, wouldn't those also belong under the `gr` prefix? E.g. `gri` for implementation, `grt` for type definition.

**Comment by zoriya** [comment]
I tried searching the issue tracker, but I'm pretty sure `grt` (for type definition) & `gd`/`gD` (for declaration/definitions) never got mapped. Should I open a PR for it?

**Comment by iwinux** [comment]
Why is `vim.lsp.buf.definition` left out in https://github.com/neovim/neovim/pull/28650? It looks very inconsistent - why not `grd` for `vim.lsp.buf.definition`?

**Comment by justinmk** [comment]
> It looks very inconsistent - why not `grd` for `vim.lsp.buf.definition`?

because `gd` and `ctrl-]` literally already exist for that purpose.
